Long explainer checkbox

shtukas opened the pull request:
This introduces the 'Long explainer' checkbox allowing users to specify if an explainer is
Expandable or Flat. Note that this currently doesn't affect the workings of the Editor itself,
but the information is recorded.

Testing:
- run the code
- click on the checkbox
- reload the page. The checkbox should be in "checked" state.
- click on the checkbox
- reload the page. The checkbox should be in "un-checked" state.

shtukas committed:
Long explainer checkbox

This introduces the 'Long explainer' checkbox allowing users to specify if an explainer is
Expandable or Flat. Note that this currently doesn't affect the workings of the Editor itself,
but the information is recorded.

Testing:
- run the code
- click on the checkbox
- reload the page. The checkbox should be in "checked" state.
- click on the checkbox
- reload the page. The checkbox should be in "un-checked" state.

shtukas commented:
![picture 1](https://cloud.githubusercontent.com/assets/6035518/17098492/94b64ed6-525a-11e6-8c89-dfba1ec5ca77.png)

shtukas commented:
Note: compiles and runs fine on local

philmcmahon commented on explainer-server/app/views/explainEditor.scala.html:
```diff
@@ -68,6 +66,24 @@
+            var checkedstring = displayType=='Expandable' ? ' checked ' : ''
+            $(".explainer-editor__displayType-inner").html('<input class="explainer-editor__displayType-checkbox" type="checkbox" name="displayType" value="long" '+checkedstring+'> Long explainer')
+        }
+        $(document).delegate(".explainer-editor__displayType-checkbox", 'click', function() {
```
could you use [onchange](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Attribute/onchange) to simplify this line?

shtukas replied:
`delegate` allows setting an event on a DOM element that has not been rendered yet (reason why I used it), but I can easily change `click` to `change`. Now it is 

```
        $(document).delegate(".explainer-editor__displayType-checkbox",'change',function() {
            var $this = $(this);
            if ($this.is(':checked')) {
                ExplainEditorJS().setDisplayType("@id","Expandable")
            } else {
                ExplainEditorJS().setDisplayType("@id","Flat")
            }
        });
```

works as expected

philmcmahon commented on explainer-server/app/views/explainEditor.scala.html:
```diff
@@ -68,6 +66,24 @@
+        function installDisplayTypeTickBox(displayType){
+            console.log(displayType);
+            var checkedstring = displayType=='Expandable' ? ' checked ' : ''
+            $(".explainer-editor__displayType-inner").html('<input class="explainer-editor__displayType-checkbox" type="checkbox" name="displayType" value="long" '+checkedstring+'> Long explainer')
```
I'm still kinda sad that we're writing HTML like this. Could we sit together and try find a way to write this with ScalaTags? (Where you'll probably just prove to me that it's impossible but at least I'll learn something)

philmcmahon replied:
Also it's best not to use class selectors when we want to pick out an individual element - better to give the div an ID and use that.

shaundillon replied:
@philmcmahon Scala tags would be best if possible, but, if not, there's a more JS-y way of doing this using the DOM api ([this sort of stuff](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement))

Also, for your second one, IDs in HTML just lead to pain and couples the JS a bit too tightly to the HTML without any real benefits I feel. I like [data attributes](https://developer.mozilla.org/en/docs/Web/Guide/HTML/Using_data_attributes) or, failing that, a class as the JS shouldn't care whether it's unique or not.

shtukas replied:
re selector: @ShaunYearStrong spent all of last week teaching me to avoid ids even for unique elements (css best practices)

shtukas replied:
@philmcmahon @ShaunYearStrong 

So I would like to sort of make a move on this...

**javascript versus scalatags**

Generating the input on scalaTags, is possible but the event handler then suffers from the same caching problem as before. In other words the following caches the state of the checkbox

```
    val checkbox: TypedTag[Input] = input(
      id:="mychecbbox",
      `type`:="checkbox"
    )
    val check = checkbox.render
    check.onchange = (e: dom.Event) => {
      g.console.log( check.value )
    }
```

So I will stick use scalaTags to generate the input but will use javascript to handle the event. 

**CSS classes versus ids**

Will follow CSS best practices. So I will use classes (as currently the case).

**DOM API**

Thanks for the suggestion. I will do that next time I need to generate DOM elements. I actually thought about it, but then didn't see a big deal with injecting the HTML. Bad Pascal... 

Cheers,

ps: see coming commit ...

shaundillon replied:
@shtukas - Just to clarify on the classes point. It's 100% best practice to use classes for everything in CSS, in JS it's more of a "why not?" kind of case. The uniqueness denoted by an ID should only apply to the HTML. For JS you're better off using the data attributes or selecting some other way (creating the element in JS and binding it there, for example).

shtukas replied:
okey-dokey! Cheers,

philmcmahon replied:
Sounds good!

shtukas committed:
-

shtukas committed:
Code refactoring

In this commit

- Generate the input element using scalatag.
- Change the event that the checkbox listens to.

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -138,11 +140,33 @@ object ExplainEditorJS {
+    val checkbox: TypedTag[Input] = explainer.data.displayType match {
+      case "Expandable" => {
+        input(
+          cls:="explainer-editor__displayType-checkbox",
```
maybe store this in a val to avoid repeating it

shtukas replied:
Done!

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -138,11 +140,33 @@ object ExplainEditorJS {
       hr,
+      div(cls:="explainer-editor__displayType-wrapper")(
+        div(cls:="explainer-editor__displayType-inner")(
+          checkbox.render, " Long explainer"
```
maybe 'Expandable explainer'

shtukas replied:
Done!

philmcmahon commented on explainer-server/app/views/explainEditor.scala.html:
```diff
@@ -68,6 +66,21 @@
+         * This function was put here at opposition to within the scala code because
+         * the read operation of the text input uses cached values in the JS code generated by ScalaJS.
+         */
+        $(document).delegate(".explainer-editor__displayType-checkbox",'change',function() {
```
Maybe do https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Attribute/onchange instead to call this function

shtukas replied:
In this case I just let scalaTags do it for me, 

```
    val checkboxTag = checkbox.render
    checkboxTag.onchange = (x: Event) => {
      g.updateCheckboxState()
    }
```

where `updateCheckboxState()` is the JavaScript function.

shtukas committed:
small refactorings

philmcmahon commented:
👍
